### ISSUE TITLE
Topicset @type attribute warning in move-meta-entries

### ISSUE THREAD

**Issue by shaneataylor** [dita-ot/dita-ot#2321]
Version: DITA-OT 2.2.4
Transformation: HTML5 customization
Platform: OS X

Since upgrading my plugin to 2.2.4, I see warnings in my build log for topicsets:

> file:/Users/staylor/Dev/dita_source_files/m_a_common_topicsets.ditamap:12:69: [DOTX019W][WARN]: The type attribute on a topicref was set to 'topicset', but the topicref references a 'concept' topic. This may cause your links to sort incorrectly in the output. Note that the type attribute cascades in maps, so the value 'topicset' may come from an ancestor topicref.

My map doesn't specify `@type="topicset"` anywhere. I'm reasonably sure this is an error in xsl/preprocess/mappullImpl.xsl:

```
<xsl:template match="*" mode="mappull:verify-type-value">
    <xsl:param name="type" as="xs:string"/>          <!-- Specified type on the topicref -->
    <xsl:param name="actual-class" as="xs:string"/>  <!-- Class value on the target element -->
    <xsl:param name="actual-name" as="xs:string"/>   <!-- Name of the target element -->
    <xsl:param name="WORKDIR" as="xs:string">
      <xsl:apply-templates select="/processing-instruction('workdir-uri')[1]" mode="get-work-dir"/>
    </xsl:param>
    <xsl:choose>
      <!-- The type is correct; concept typed as concept, newtype defined as newtype -->
      <xsl:when test="$type=$actual-name"/>
      <!-- If the actual class contains the specified type; reference can be called topic,
           specializedReference can be called reference -->
      <xsl:when test="contains($actual-class,concat(' ',$type,'/',$type,' '))">
        <!-- commented out for bug:1771123 start -->
        <!--xsl:apply-templates select="." mode="ditamsg:type-mismatch-info">
          <xsl:with-param name="type" select="$type"/>
          <xsl:with-param name="actual-name" select="$actual-name"/>
        </xsl:apply-templates-->
        <!-- commented out for bug:1771123 end -->
      </xsl:when>
      <!-- Otherwise: incorrect type is specified -->
      <xsl:otherwise>
        <xsl:apply-templates select="." mode="ditamsg:type-mismatch-warning">
          <xsl:with-param name="type" select="$type"/>
          <xsl:with-param name="actual-name" select="$actual-name"/>
        </xsl:apply-templates>
      </xsl:otherwise>
    </xsl:choose>
  </xsl:template>
```

It looks to me like when you use a `<topicsetref>`, the processing sees the target element name as topicset and that results in a mismatch. Wondering if the fix is simply to ignore this check when the target is a topicset.

**Comment by drmacro** [comment]
The @type attribute for topicsetref is set to "topicset" by default so if you want to refer to a different topicref type you'll need to set it explicitly to match the target type. Or don't use topicsetref.

**Comment by robander** [comment]
I'd go with the "don't use `<topicsetref>`" option myself. At this point I rather hope it will be deprecated by OASIS, but that's up to the TC to decide....

**Comment by robander** [comment]
But yeah, that logic is based on the assumption that `@type` will refer to a topic element - thus the check for actual matching element name, followed by `value/value` in the `@class` attribute. This assumes that the target is a structural topic specialization (like when `<task>` element or specialization has `task/task` in the class attribute). It looks like that first check for element name should allow `<topicsetref>` to work when it refers to a `<topicset>`, but throw the warning for any other map element.

**Comment by robander** [comment]
In your source content, is the `<topicsetref>` element actually referencing a `<topicset>`? 
* If not, I think the warning is valid. 
* If so, then I think we need a more general fix, because any `<topicref>` specialization that references a `<topicset>` with `type="topicset"` should not throw out this error.

**Comment by robander** [comment]
Figured out the root of the issue.

The `<topicsetref>` element is resolved by `mapref`, which takes care of references to other maps (or partial maps).

I created one that refers to a `<topicset>`, but after resolution - just like conref - the local `@type` attribute on the references is preserved, resulting in this resolved element:
```xml
<topicset class="+ map/topicref mapgroup-d/topicsetref " id="wrap" type="topicset">
```

The `@class` attribute is odd there too, because of kind of the complex / sometimes unpleasant resolution described in the spec for map references. Because `@class` is preserved, the code fix in pull request #2322 does work, but I think there is really a larger issue. 

`@type` is defined as follows: "[The @type attribute is used on cross-references to describe the target of the reference.](http://docs.oasis-open.org/dita/dita/v1.3/os/part1-base/archSpec/base/dita-linking.html#dita-linking)" So, when referencing a map or map element, it is proper to use `@type` to describe that element. By preserving that after the reference is resolved, we end up with a `@type` value that described the original reference, but now *appears* to describe something else - and cascades down further within the resolved branch.

I think that the proper fix here is to update `mapref` so that it drops `@type` as part of the resolution. Does anybody see a problem with that? I don't think we are validating `@type` on that reference today, but if it is used properly, I can't think of a case where it would be correct *after* resolving the map reference.

**Comment by robander** [comment]
Sample files I used in my test - 2321.ditamap:

```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE map PUBLIC "-//OASIS//DTD DITA Map//EN"
 "../dtd/technicalContent/dtd/map.dtd">
<map>
  <topicset id="wrap">
    <topicref href="2321.dita" id="simple"/>
  </topicset>
  <topicsetref href="#wrap"/>
</map>
```

2321.dita:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE concept PUBLIC "-//OASIS//DTD DITA Concept//EN" "concept.dtd">
<concept id="concept_f3w_rnp_tv">
 <title>Sample topic in a topicset</title>
 <shortdesc>for testing</shortdesc>
 <conbody/>
</concept>
```

**Comment by shaneataylor** [comment]
@robander I appreciate your digging into this. Your approach makes sense to me.

**Comment by robander** [comment]
And finally - just to illustrate why this should be fixed in `mapref` rather than in later stages - you'll get the same warnings if you use the following markup in a bookmap:
```xml
<chapter id="mychap">...</chapter>
...
<chapter href="#mychap" type="chapter" format="ditamap"/>
```

**Comment by drmacro** [comment]
Definitely maintaining the original @type and @class makes no sense--the resolved result should reflect the target.

**Comment by robander** [comment]
@shaneataylor - cool - I'll mark the existing pull request closed and we'll see about a better fix.

@drmacro - the `@class` part is an ugly aspect of the resolution required in many cases - pretty sure we've discussed elsewhere and nobody really likes it. Another thing on my list of stuff to improve in DITA 2.0.

**Comment by robander** [comment]
Looks like adding `@type` to the end of this line fixes the problem:
https://github.com/dita-ot/dita-ot/blob/develop/src/main/xsl/preprocess/maprefImpl.xsl#L168

```
 <xsl:apply-templates select="@* except (@class, @href, @dita-ot:orig-href, @format, @dita-ot:orig-format, @keys, @keyscope, @type)"/>
```

**Comment by robander** [comment]
We discussed `<topicsetref>` at today's OASIS TC meeting. I suggested removing it for DITA 2.0. For whatever that's worth... :innocent:

**Comment by robander** [comment]
Fixed with pull request #2338
